wasm.md

```diff
@@ -1186,22 +1284,82 @@ Start Function
     syntax Instr ::= "(" "start" TextFormatIdx ")"
  // ----------------------------------------------
     rule <k> ( start TFIDX ) => ( invoke FADDR ) ... </k>
-         <funcIds> IDS </funcIds>
-         <funcIndices> ... #ContextLookup(IDS , TFIDX) |-> FADDR ... </funcIndices>
+         <curModIdx> CUR </curModIdx>
+         <moduleInst>
+           <modIdx> CUR </modIdx>
+           <funcIds> IDS </funcIds>
+           <funcIndices> ... #ContextLookup(IDS , TFIDX) |-> FADDR ... </funcIndices>
+           ...
+         </moduleInst>
 ```
 
-Module Declaration
-------------------
+Module Instantiation
+--------------------
 
-Currently, we support a single module.
-The surronding `module` tag is discarded, and the inner portions are run like they are instructions.
+A new module instance gets allocated.
+Then, the surrounding `module` tag is discarded, and the definitions are executed, putting them into the module currently being defined.
 
 ```k
     syntax Stmt ::= "(" "module" Defns ")"
- // --------------------------------------
+                  | "(" "module" Identifier Defns ")"
+ // -------------------------------------------------
+ // Map the given id to the next available module number, and then define the module as usual.
+    rule <k> ( module ID:Identifier DEFNS ) => ( module DEFNS ) ... </k>
+         <moduleIds> ... .Map => ID |-> NEXT ... </moduleIds>
+         <nextModuleIdx> NEXT </nextModuleIdx>
     rule <k> ( module DEFNS ) => DEFNS ... </k>
+         <curModIdx> _ => NEXT </curModIdx>
+         <nextModuleIdx> NEXT => NEXT +Int 1 </nextModuleIdx>
+         <moduleInstances>
+           (.Bag =>
+             <moduleInst>
+               <modIdx>        NEXT    </modIdx>
+               <typeIds>       .Map </typeIds>
+               <funcIds>       .Map </funcIds>
+               <nextTypeIdx>   0    </nextTypeIdx>
+               <nextFuncIdx>   0    </nextFuncIdx>
+               <nextGlobalIdx> 0    </nextGlobalIdx>
+               <types>         .Map </types>
+               <funcIndices>   .Map </funcIndices>
+               <tabIndices>    .Map </tabIndices>
+               <memIndices>    .Map </memIndices>
+               <globalIndices> .Map </globalIndices>
+               <exports>       .Map </exports>
+             </moduleInst>
+           )
+           ...
+         </moduleInstances>
 ```
 
+Note that we want to always have a module available, for execution Statements outside of modules.
+We do this with a small hack to ensure that a first module is always created when needed, if not available.
+
+```k
+    rule <curModIdx> 0 </curModIdx>
+         <nextModuleIdx> 0 => 1 </nextModuleIdx>
+         <moduleInstances>
+           (.Bag =>
+             <moduleInst>
```

Comment:
same formatting here.

Also, I think you can just do:

```
  ( .Bag
 => <moduleInst>
      <modIdx> 0 </modIdx>
      ...
    </moduleInst>
  )
```

and the remaining cells will get default values from the `configuration` declaration, which makes it more robust to changes in the `configuration`.